syntax = "proto3";

package silverbullet_rag;

// RAG service for fast hooks integration
service RAGService {
  // Query the knowledge graph
  rpc Query(QueryRequest) returns (QueryResponse);

  // Search by keyword
  rpc Search(SearchRequest) returns (SearchResponse);

  // Semantic search using vector embeddings
  rpc SemanticSearch(SemanticSearchRequest) returns (SemanticSearchResponse);

  // Hybrid search combining keyword and semantic search
  rpc HybridSearch(HybridSearchRequest) returns (HybridSearchResponse);

  // Read a page from the space
  rpc ReadPage(ReadPageRequest) returns (ReadPageResponse);

  // Propose a change to a page (creates a proposal for user review)
  rpc ProposeChange(ProposeChangeRequest) returns (ProposeChangeResponse);

  // List change proposals by status
  rpc ListProposals(ListProposalsRequest) returns (ListProposalsResponse);

  // Withdraw a pending proposal
  rpc WithdrawProposal(WithdrawProposalRequest) returns (WithdrawProposalResponse);

  // Get context for an Open WebUI folder (finds pages with matching openwebui-folder frontmatter)
  rpc GetFolderContext(GetFolderContextRequest) returns (GetFolderContextResponse);

  // Get project context by GitHub remote or folder path
  rpc GetProjectContext(GetProjectContextRequest) returns (GetProjectContextResponse);
}

message QueryRequest {
  string cypher_query = 1;
}

message QueryResponse {
  string results_json = 1;
  bool success = 2;
  string error = 3;
}

message SearchRequest {
  string keyword = 1;
  int32 limit = 2;
}

message SearchResponse {
  string results_json = 1;
  bool success = 2;
  string error = 3;
}

message SemanticSearchRequest {
  string query = 1;
  int32 limit = 2;
  repeated string filter_tags = 3;
  repeated string filter_pages = 4;
}

message SemanticSearchResponse {
  string results_json = 1;
  bool success = 2;
  string error = 3;
}

message HybridSearchRequest {
  string query = 1;
  int32 limit = 2;
  repeated string filter_tags = 3;
  repeated string filter_pages = 4;
  string fusion_method = 5;  // "rrf" or "weighted"
  float semantic_weight = 6;
  float keyword_weight = 7;
}

message HybridSearchResponse {
  string results_json = 1;
  bool success = 2;
  string error = 3;
}

// ReadPage messages
message ReadPageRequest {
  string page_name = 1;
}

message ReadPageResponse {
  bool success = 1;
  string error = 2;
  string content = 3;
}

// ProposeChange messages
message ProposeChangeRequest {
  string target_page = 1;
  string content = 2;
  string title = 3;
  string description = 4;
}

message ProposeChangeResponse {
  bool success = 1;
  string error = 2;
  string proposal_path = 3;
  bool is_new_page = 4;
  string message = 5;
}

// ListProposals messages
message ListProposalsRequest {
  string status = 1;  // "pending", "accepted", "rejected", or "all"
}

message ProposalInfo {
  string path = 1;
  string target_page = 2;
  string title = 3;
  string description = 4;
  string status = 5;
  bool is_new_page = 6;
  string proposed_by = 7;
  string created_at = 8;
}

message ListProposalsResponse {
  bool success = 1;
  string error = 2;
  int32 count = 3;
  repeated ProposalInfo proposals = 4;
}

// WithdrawProposal messages
message WithdrawProposalRequest {
  string proposal_path = 1;
}

message WithdrawProposalResponse {
  bool success = 1;
  string error = 2;
  string message = 3;
}

// GetFolderContext messages
message GetFolderContextRequest {
  string folder_path = 1;  // Open WebUI folder path (e.g., "Projects/MyProject")
}

message GetFolderContextResponse {
  bool success = 1;
  string error = 2;
  bool found = 3;  // Whether a matching page was found
  string page_name = 4;  // Name of the matched page
  string page_content = 5;  // Full content of the matched page
  string folder_scope = 6;  // Silverbullet folder to scope searches to
}

// GetProjectContext messages
message GetProjectContextRequest {
  string github_remote = 1;  // GitHub repository in "owner/repo" format
  string folder_path = 2;    // Folder path in Silverbullet space (e.g., "Projects/MyProject")
}

message RelatedPage {
  string name = 1;
  string path = 2;
}

message ProjectInfo {
  string file = 1;           // Relative path to project file
  string github = 2;         // GitHub remote if set
  repeated string tags = 3;  // Tags from frontmatter
  repeated string concerns = 4;  // Concerns from frontmatter
  string content = 5;        // Page content (frontmatter stripped)
}

message GetProjectContextResponse {
  bool success = 1;
  string error = 2;
  ProjectInfo project = 3;
  repeated RelatedPage related_pages = 4;
}
